design/ceph/multus-network.md

```diff
@@ -0,0 +1,129 @@
+
+You can find a more detailed analysis at the end of the document in the [rejected proposal section](#rejected-proposals).
+
+## Ceph daemons implementation challenges
+
+### Monitors and OSDs
+
+The Ceph monitors only need access to the public network.
+The OSDs needs access to both public and cluster networks.
+
+Monitors requirements so far are the following:
+
+- Predictable IP addresses
+- Keep the same IP address for the life time of the deployment (IP should survive a restart)
+
+### RGW implementation
+
+Only need access to the public network.
+They use service IP with a load-balancer so we need to be careful.
+
+### MDS/RBD-MIRROR/NFS implementation
+
+Only need access to the public network.
+Nothing to do in particular since they don't use any service IPs.
+
+### CSI pods
+
+We can add annotations to these pods and they can reach out the Ceph public network, then the driver will expose the block or the filesystem normally.
+
+## Accepted proposal
+
+So far, the team has decided to go with the [whereabouts](https://github.com/dougbtv/whereabouts) IPAM.
+It's an IP Address Management (IPAM) CNI plugin that assigns IP addresses cluster-wide.
+If you need a way to assign IP addresses dynamically across your cluster -- Whereabouts is the tool for you. If you've found that you like how the host-local CNI plugin works, but, you need something that works across all the nodes in your cluster (host-local only knows how to assign IPs to pods on the same node) -- Whereabouts is just what you're looking for.
+Whereabouts can be used for both IPv4 & IPv6 addressing.
+
+It is under active development and is not ready but ultimately will allow to:
+
+- Have static IP addresses distributed across the cluster
+- These IPs will survive any deployment restart
+- The allocation will be done by whereabouts and not by Rook
+
+## Rejected proposals
+
+The following proposal were rejected but we keep them here for traceability and knowledge.
+
+### IPAM type 'DHCP'
+
+In this scenario, a DHCP server will distribute an IP address to a pod using a given range.
+
+Pros:
+
+- Pods will get a dedicated IP on a physical network interface
+- No changes required in Ceph, Rook will detect the CIDR via the `NetworkAttachmentDefinition` then populate Ceph's flag `public_network` and `cluster_network`
+
+Cons:
+
+- IP allocation not predictable, we don't know it until the pod is up and running.
+So the detection must append inside the monitor container is running, similarly to what the OSD code does today.
+- Requires drastic changes in the monitor bootstrap code
+- This adds a DHCP daemon on every part of the cluster and this has proven to be troublesome
+
+Assuming we go with this solution, we might need to change the way the monitors are bootstrapped:
+
+1. let the monitor discovered its own IP based on an interface
+2. once the first mon is bootstrapped, we register its IP in a ConfigMap as well as populating clusterInfo
```

Comment:
Can the operator discover this ip address and then update the configmap? Or do we need something in the mon pod to discover it? None of us want to add the rook image back into the mon pods though.

Reply:
We don't know the IP address until the pod is up and running so the operator cannot discover it. And yes, we don't want to use the Rook binary anymore.